gnd/src/main/java/com/google/android/gnd/persistence/local/room/RoomLocalDataStore.java

```diff
@@ -298,22 +299,68 @@ public Completable mergeObservation(Observation observation) {
     ObservationEntity observationEntity = ObservationEntity.fromObservation(observation);
     return observationMutationDao
         .findByObservationId(observation.getId())
-        .map(observationEntity::applyMutations)
-        .flatMapCompletable(observationDao::insertOrUpdate)
+        .flatMapCompletable(mutations -> mergeObservation(observationEntity, mutations));
+  }
+
+  private Completable mergeObservation(
+      ObservationEntity observation, List<ObservationMutationEntity> mutations) {
+    if (mutations.isEmpty()) {
+      return Completable.complete();
+    }
+    ObservationMutationEntity lastMutation = mutations.get(mutations.size() - 1);
+    return loadUser(lastMutation.getUserId())
+        .map(user -> applyMutations(observation, mutations, user))
+        .flatMapCompletable(obs -> observationDao.insertOrUpdate(obs).subscribeOn(Schedulers.io()));
+  }
+
+  private ObservationEntity applyMutations(
+      ObservationEntity observation, List<ObservationMutationEntity> mutations, User user) {
+    ObservationMutationEntity lastMutation = mutations.get(mutations.size() - 1);
+    long clientTimestamp = lastMutation.getClientTimestamp();
+    Log.v(TAG, "Merging observation " + this + " with mutations " + mutations);
+    ObservationEntity.Builder builder = observation.toBuilder();
+    // Merge changes to responses.
+    for (ObservationMutationEntity mutation : mutations) {
+      builder.applyMutation(mutation);
+    }
+    // Update modified user and time.
+    AuditInfoEntity lastModified =
+        AuditInfoEntity.builder()
+            .setUser(UserDetails.fromUser(user))
+            .setClientTimeMillis(clientTimestamp)
+            .build();
+    builder.setLastModified(lastModified);
+    Log.v(TAG, "Merged observation " + builder.build());
+    return builder.build();
+  }
+
+  @Override
+  public Single<User> loadUser(String id) {
+    return userDao
+        .findById(id)
+        .doOnComplete(() -> Log.e(TAG, "User missing local db: " + id))
```

Comment:
Would this trigger even when the user is found, since it is within doOnComplete?

Reply:
You have an impeccable eye for detail! It should have been `doOnError` Fixed.

Added some follow-up TODOs in the description of #15.